### ISSUE TITLE
Remove repr() and getDelayedCalls() usage from twisted.web.test.test_web.SessionTest.test_delayedCallCleanup

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1996]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1996|
|Type|task|
|Created|2006-08-13 22:06:33Z|

Comparing repr()s of things is pretty terrible.  We shouldn't do it.  Structured objects FTW.

getDelayedCalls should never have been added to IReactorTime.  We should avoid using it wherever possible, remove all existing usages of it from the codebase, and then remove it.  But that's a separate ticket.  For this ticket, just remove it from test_delayedCallCleanup.

Session creation should be parameterized such that the clock used for scheduling is accessible to the test suite and a fake one which does not involve the reactor or real wallclock time can be specified for the test.

Attachments:

* [web_1996.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/204/204d1b68ca70c70e17417076588df954f47da0da/e36d27ea39d8d6d460c4f200b66b379ecd0965a2.diff) (8749 bytes) - added by therve on 2007-01-12 15:06:06Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1996 1996
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1155506793000000 1155506793000000
changetime__1169198417000000 1169198417000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I started to do this but the branch was merged before I finished:

```
Index: twisted/web/server.py
===================================================================
--- twisted/web/server.py       (revision 17814)
+++ twisted/web/server.py       (working copy)
@@ -399,10 +399,15 @@
         self.site = site
         self.uid = uid
         self.expireCallbacks = []
-        self.checkExpiredLoop = task.LoopingCall(self.checkExpired)
         self.touch()
         self.sessionNamespaces = {}
 
+
+    def startCheckingExpiration(self, lifetime):
+        self.checkExpiredLoop = task.LoopingCall(self.checkExpired)
+        self.checkExpiredLoop.start(lifetime)
+
+
     def notifyOnExpire(self, callback):
         """Call this callback when the session expires or logs out.
         """
@@ -443,6 +448,7 @@
 
     counter = 0
     requestFactory = Request
+    sessionFactory = Session
     displayTracebacks = True
     
     def __init__(self, resource, logPath=None, timeout=60*60*12):
@@ -472,8 +478,8 @@
         """Generate a new Session instance, and store it for future reference.
         """
         uid = self._mkuid()
-        session = self.sessions[uid] = Session(self, uid)
-        session.checkExpiredLoop.start(1800)
+        session = self.sessions[uid] = self.sessionFactory(self, uid)
+        session.startCheckingExpiration(1800)
         return session
 
     def getSession(self, uid):
```

**Comment by twisted-trac** [comment]
|[<img alt="keturn's avatar" src="https://avatars.githubusercontent.com/u/83819?s=50" width="50" height="50">](https://github.com/keturn)<a name="note_2"></a>|@keturn commented|
|-|-|

I won't say that getDelayedCalls is not a bad thing to use, but I fear what you're suggesting for test_delayedCallCleanup changes the meaning of the test.  That particular test is to ensure that the implementation of "startCheckingExpiration" that people use does not leave crap in the scheduler's reactor after the Session goes away.  If you use a different implementation of startCheckingExpiration in the test, then you aren't testing that.

I also haven't figured out a way to test that without the use of getDelayedCalls().

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

I went on and wrote something. acapnotic is right that we can't really test if there is no remain DelayedCall somewhere, but I considered that checking loop is enough. I also extracted some constants as class variables.

FWIW, I consider having a LoopingCall by session bad design. We should have a looping call for the Site checking all the sessions, unless I miss something.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

Changes look good, for the most part.  A couple very minor points:

  * Instead of ```if self.checkExpiredLoop:```, consider ```if self.checkExpiredLoop is not None:```.  The case being tested for is really that a LoopingCall has been assigned to the attribute, not anything about the boolean value of the object it is bound to.  Checking against ```None``` explicitly more directly expresses the intent here, in addition to avoiding all of the ```__nonzero__``` stuff Python does for the former spelling.
  * ```@ivar``` doc markup for the new attributes on Session and Site would be good.
  * There are a few pyflakes warnings for test_web.py.  As long as the file is being modified, it'd be good to clean these up.

The parameterization of the session timeout which is done here looks like it should also resolve [#10717](https://github.com/twisted/twisted/issues/10717), as I suspect you have already noticed. :)  If you agree, resolve that ticket as well when you merge this branch.

Also, thanks for the docstring reformatting.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Putting it back for review(in session-cleanup-1996), for my far from perfect english, as I added a few doctrings. I also cleanups test_web and removed a useless string import.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Looks good.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [19411]) Merge session-cleanup-1996

Author: therve
Reviewer: exarkun
Fixes [#1996](https://github.com/twisted/twisted/issues/1996)

getDelayedCalls() was used in twisted.web.test.test_web.SessionTest, whereas
it's an internal method that should be removed. The test has been rewritten, 
and, in the process, web.Session and web.Site have been refactored to be tested
easier. [#10717](https://github.com/twisted/twisted/issues/10717) is also fixed by the changes.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
